### PR TITLE
feat: Unfold widget

### PR DESCRIPTION
**Pull request by PatrickMassot**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Widget/Unfold.lean [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2023 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot
```
Comment (YaelDillies) [inline Mathlib/Tactic/Widget/Unfold.lean:4]: This is taken from Jovan's code, right?

File: Mathlib/Tactic/Widget/Unfold.lean [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+This file defines a `unfold?` tactic that displays a widget panel allowing to generate
+an `unfold` call with name specified by selecting a subexpression in the infoview.
+-/
+
+open Lean Meta Server ProofWidgets
+
+/-- If the given expression is a constant, return its name, if it is an application of a constant,
+returns its name, otherwise return `none`. -/
+def Lean.Expr.headName : Expr → Option Name
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/Unfold.lean:18]:
```suggestion
def Lean.Expr.headName? : Expr → Option Name
```
or maybe
```suggestion
def Lean.Expr.headConstName? : Expr → Option Name
```
to match `Lean.Expr.constName?`

File: Mathlib/Tactic/Widget/Unfold.lean [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  let some constName := (← Lean.Core.viewSubexpr subexprPos tgtExpr).headName
+    | throwError "You must select a name."
+  pure (s!"unfold {constName}{loc}", s!"unfold {constName}{loc}", none)
+
+/-- Rpc function for the conv widget. -/
+@[server_rpc_method]
+def UnfoldSelectionPanel.rpc :=
+mkSelectionPanelRPC insertUnfold
+  "Use shift-click to select one constant that you want to unfold."
+  "Unfold 🔍" (onlyGoal := false) (onlyOne := true)
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/Unfold.lean:45]:
```suggestion
  mkSelectionPanelRPC insertUnfold
    "Use shift-click to select one constant that you want to unfold."
    "Unfold 🔍" (onlyGoal := false) (onlyOne := true)
```

File: Mathlib/Tactic/Widget/Unfold.lean [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  | ⟨_, .target subexprPos⟩ => pure (none, subexprPos)
+  | ⟨_, .hypType fvar subexprPos⟩ => pure (some fvar, subexprPos)
+  | ⟨_, .hypValue fvar subexprPos⟩ => pure (some fvar, subexprPos)
+  | _ => throwError "You must select something in the goal or in a local value."
+  let (tgtExpr, loc) ← match fvar with
+  | some fvarId => pure (← fvarId.getType, s!" at {← fvarId.getUserName} ")
+  | none => pure (goalType, "")
+  let some constName := (← Lean.Core.viewSubexpr subexprPos tgtExpr).headName
+    | throwError "You must select a name."
+  pure (s!"unfold {constName}{loc}", s!"unfold {constName}{loc}", none)
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/Unfold.lean:38]: Is this unrelated to the "try this" framework? It's not immediately obvious to me that I'm expected to click on the `unfold foo` that appears.

File: Mathlib/Tactic/Widget/Unfold.lean [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+def Lean.Expr.headName : Expr → Option Name
+  | .const n .. => n
+  | .app fn _ => fn.headName
+  | _ => none
+
+/-- Return the link text and inserted text for the unfold widget. -/
+@[nolint unusedArguments]
+def insertUnfold (locations : Array Lean.SubExpr.GoalsLocation) (goalType : Expr)
+    (_ : SelectInsertParams) : MetaM (String × String × Option (String.Pos × String.Pos)) := do
+  let some pos := locations[0]? | throwError "You must select something."
```
Comment (eric-wieser) [inline Mathlib/Tactic/Widget/Unfold.lean:27]: Is there any reason to restrict to one expression?

File: Mathlib/Tactic/Widget/Unfold.lean [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  | ⟨_, .hypValue fvar subexprPos⟩ => pure (some fvar, subexprPos)
+  | _ => throwError "You must select something in the goal or in a local value."
+  let (tgtExpr, loc) ← match fvar with
+  | some fvarId => pure (← fvarId.getType, s!" at {← fvarId.getUserName} ")
+  | none => pure (goalType, "")
+  let some constName := (← Lean.Core.viewSubexpr subexprPos tgtExpr).headName
+    | throwError "You must select a name."
+  pure (s!"unfold {constName}{loc}", s!"unfold {constName}{loc}", none)
+
+/-- Rpc function for the conv widget. -/
```
Comment (alexjbest) [inline Mathlib/Tactic/Widget/Unfold.lean:40]:
```suggestion
/-- Rpc function for the unfold widget. -/
```
perhaps?

File: Mathlib/Tactic/Widget/Unfold.lean (hunk 2 above)
Comment (alexjbest) [inline Mathlib/Tactic/Widget/Unfold.lean:18]: I'm also not sure what the good meta style should be here, I think this is the same as doing `e.getAppFn.constName?` (or even `e.getAppFun'.constName?` which is probably a good idea here, it ignores mdata). So, do we need a new def, and if we do should it be defined recursively like this or via the existing ones directly.
I have a vague feeling that reusing existing components would be better in some cases for code generation (smaller dlls, more chance that someone adds a fast `implemented_by` / `extern` and it helps you) but no actual data there.

**Comment by fpvandoorn** [comment]
please mark as `awaiting-review` if it's ready for review

**Comment by PatrickMassot** [comment]
I hope this is superseded by #11768.

**Pull request closed without merging**